torch/csrc/autograd/variable.cpp

```diff
@@ -145,56 +77,50 @@ void Variable::backward(
 
   std::vector<Variable> inputs;
   if (!gradient.has_value()) {
-    gradient = make_variable(at::ones_like(data()), /*requires_grad=*/false);
+    gradient = at::ones_like(*this);
+    as_variable_ref(*gradient).set_requires_grad(false);
   }
   inputs.push_back(std::move(as_variable_ref(*gradient)));
   Engine::get_default_engine().execute(edges, inputs, keep_graph, create_graph);
 }
 
-void Variable::Impl::set_data(const at::Tensor &new_data) {
+void Variable::set_data(const at::Tensor &new_data) {
+  // `var.set_data(new_data)` shallow-copies all non-autograd TensorImpl fields
+  // from `new_data` to `var`. It requires that `new_data` has the same derived
+  // type of TensorImpl as `var`.
+  TORCH_CHECK(
+    typeid(*this->unsafeGetTensorImpl()) == typeid(*new_data.unsafeGetTensorImpl()),
```

Comment:
nit: can you make the operator precedence clear here, i.e.
`typeid(*(new_data.unsafeGetTensorImpl()))`

Reply:
also, do you have a test at the python level for this?